registry/storage/driver/middleware/cloudfront/middleware.go

```diff
@@ -90,23 +90,22 @@ func newCloudFrontStorageMiddleware(storageDriver storagedriver.StorageDriver, o
 	return &cloudFrontStorageMiddleware{StorageDriver: storageDriver, cloudfront: cf, duration: duration}, nil
 }
 
+// cloudFrontPather is any type that is capable of returning a *relative* path
+// which may be cached by CloudFront.
+type cloudFrontPather interface {
+	CloudFrontPath(path string) string
+}
+
 // Resolve returns an http.Handler which can serve the contents of the given
 // Layer, or an error if not supported by the storagedriver.
 func (lh *cloudFrontStorageMiddleware) URLFor(path string, options map[string]interface{}) (string, error) {
 	// TODO(endophage): currently only supports S3
-	options["expiry"] = time.Now().Add(lh.duration)
-
-	layerURLStr, err := lh.StorageDriver.URLFor(path, options)
-	if err != nil {
-		return "", err
-	}
-
-	layerURL, err := url.Parse(layerURLStr)
-	if err != nil {
-		return "", err
+	pather, ok := lh.StorageDriver.(cloudFrontPather)
+	if !ok {
+		return "", fmt.Errorf("the CloudFront middleware does not support this backend storage driver")
```

Comment:
This should probably be `storagedriver.ErrUnsupportedMethod`. Add warning level message that indicates that the configuration is likely bad.